src/OpenIddict.Core/OpenIddictController.cs

```diff
@@ -115,14 +115,20 @@ public class OpenIddictController<TUser, TApplication> : Controller where TUser
             var identity = new ClaimsIdentity(Options.AuthenticationScheme);
             identity.AddClaim(ClaimTypes.NameIdentifier, await Manager.GetUserIdAsync(user));
 
-            // Only add the name claim if the "profile" scope was present in the token request.
-            if (request.GetScopes().Contains("profile", StringComparer.OrdinalIgnoreCase)) {
-                identity.AddClaim(ClaimTypes.Name, await Manager.GetUserNameAsync(user), destination: "id_token token");
+            // Resolve the username and the email address corresponding to the user.
+            var username = await Manager.GetUserNameAsync(user);
+            var email = await Manager.GetEmailAsync(user);
+
+            // Only add the name claim if the "profile" scope was present in the token request
+            // and if the username doesn't correspond to the registered email address.
+            if (request.ContainsScope(OpenIdConnectConstants.Scopes.Profile) &&
+                !string.Equals(username, email, StringComparison.OrdinalIgnoreCase)) {
+                identity.AddClaim(ClaimTypes.Name, username, destination: "id_token token");
```

Comment:
In the default ASP.NET Identity 3 template, the username is the email address. To avoid disclosing it when the `email` scope is not requested, we ensure the username is not the same as the email address before adding it.